### DISCUSSION TITLE
Feature request: Ability to adjust bit and sample rate

### DISCUSSION THREAD

**Discussion by audiamus** [Mbucari/AAXClean.Codecs#3] [Ideas]
Will you allow me another feature request, in the course of completely eliminating the need for FFmpeg in AAX Audio Converter?

Not that I need it for myself, but people had requested that feature for AAX Audible Converter 1 several times and I implemented it with FFmpeg:

- Reduce bit and sample rate (predefined values for bit rate, with matching sample rate)
- Fixed or variable bit rate
- For both MP4/AAC and MP3 
- With progress callback

It seems that this is the only remaining feature in AAX Audio Converter 1 that cannot be be replaced with AAXClean.Codecs yet.

(AAX Audio Converter 1 also supports the old Audible Format 4, file type .aa, which is MP3, and has implemented all transcoding options for that, too, delegating to FFmpeg again. But I think, I will discontinue .aa support in AAX Audio Converter 2, since almost every user will have upgraded to .aax meanwhile.)

**Comment by Mbucari** [comment]
That feature does exist for conversion to mp3. Just use the overload that accepts an NAudio.Lame.LameConfig. You can select an combination of options you can imagine.

AAC rencoding is another matter.  I'll have to rebuild my aac libraries to support encoding.

**Reply by audiamus** [reply]
That sounds good. 

It's still a long way on my side, but planning ahead will help me. And probably not only me but you as well. From my own experience I know that reacquainting myself with a project after a couple months takes quite some time.

**Reply by Mbucari** [reply]
How are you currently transcoding AAC? FFM peg? If so, can you post the command you use here?

**Reply by audiamus** [reply]
Yes, FFmpeg.

Here is one for constant bit rate (from a log):
`-hide_banner -i "input.aax" -map_metadata -1 -map_chapters -1 -ar 22050 -b:a 64k -vn  "output.m4b"`

Not sure about variable bit rate. Have to check this in a debug environment.

**Reply by audiamus** [reply]
And here is one for variable bit rate:
`-hide_banner -i "input.aax" -map_metadata -1 -map_chapters -1 -ar 22050 -q:a 0.6 -vn  "output.m4b"`

**Comment by Mbucari** [comment]
@audiamus 

I have just successfully transcoded AAC audio with AAXClean.Codecs.  I'm using a FFmpeg version 5.1 to build the encoder, and the conversion speed with my test files is about 50x.

There's still a lot to do, but this is a huge first step.

**Reply by audiamus** [reply]
That's great. I haven't yet started in earnest on the integration of my two apps. So, still no rush here. 

_BTW, with its recent update I'm now linking directly from the AAX Audio Converter app to Book Lib Connect which is boosting the downloads for the latter. And that's downloads for AAXClean, too._

**Comment by Mbucari** [comment]
I pushed a new version with AAC support that'd I'd like you to try at some point. I added an aac encoder and completely re-wrote the decoder, and improved AAXClean's stability in general. Right now there is only single file conversion, no split support yet.

Here's an example snipet:
```C#
var options = new AacEncoderOptions
{
    BitRate = 30000,
    EncoderQuality = 0.6,
    SampleRate = SampleRate._16000,
    Stereo = false
};

await aax.ConvertToAacAsync(File.Open(@"HP_Zero_Gaiman.m4a", FileMode.OpenOrCreate, FileAccess.Write), options);
```
Note that BitRate and EncoderQuality are not mutually-exclusive options. By default, FFmpeg's aac encoder uses a quality setting of ~1.02 and a default bitrate that depends on the channel configuration.

**Reply by audiamus** [reply]
> I pushed a new version with AAC support that'd I'd like you to try at some point. 

Thank you very much, I will try it.

**Reply by Mbucari** [reply]
I also did a small rewrite of the filtering engine, so CPU utilization should be much lower for all operations.

**Reply by Mbucari** [reply]
@audiamus I pushed a new version with a new way of running async tasks. Look at AAXClean's [updated Readme](https://github.com/Mbucari/AAXClean#readme) on for an explanation and examples.

**Reply by audiamus** [reply]
Okay. I hope to find some time for a quick test the upcoming weekend.

**Reply by Mbucari** [reply]
FYI, I added multipart re-encoded mp4.

**Comment by audiamus** [comment]
I managed today to write and run a quick test. It took me a bit to get everything going, I built a debug version and integrated your github projects. Wouldn't find the native ffmpegaac dll at first. Now it's working.

My test has been CBR only so far, 22050/64k reduced to 22050/32k. Result seems as expected. Speed reported by AAXClean.Codecs was 152, compared to speed 84 with FFmpeg invoked from console, January 23 build, parameters as in my earlier post.

Both write a line to the console which appears to originate form some common code in full FFmpeg and your AAC excerpt.
[aac @ 000001de92bfcd00] Qavg: 19259.957 (AAXClean.Codecs)
[aac @ 0000023E29FF2300] Qavg: 15825.481 (FFmpeg) 

I guess a VBR test will bring no surprises.

So all my wishes are fulfilled now, many many thanks again. Now I only need to find the time to redesign AAX Audio Converter to make use of the new libs (and integrate Book Lib Connect).

**Reply by Mbucari** [reply]
> Wouldn't find the native ffmpegaac dll at first. Now it's working.

If you use the nuget package the dll files will be copied automatically. It doesn't work if you're linking directly to the source project.

> My test has been CBR only so far, 22050/64k reduced to 22050/32k. Result seems as expected. Speed reported by AAXClean.Codecs was 152, compared to speed 84 with FFmpeg invoked from console, January 23 build, parameters as in my earlier post.

Every filter in AAXClean runs on its own thread which is why it's so speedy. Deccryption happens on one thread, and decrypted frames are passed to the decoder. Decoder runs on its own thread and passes decoded frames to the encoder. Encoder runs on its own thread and writes the file.

> Both write a line to the console which appears to originate form some common code in full FFmpeg and your AAC excerpt.

Yes. It slightly annoys me, but it does show the FFmpeg roots.

> So all my wishes are fulfilled now, many many thanks again.

You're very welcome! C is a pain, but I finally got the hang of it lol.

**Reply by audiamus** [reply]
> Decryption happens on one thread, and decrypted frames are passed to the decoder. Decoder runs on its own thread and passes decoded frames to the encoder. Encoder runs on its own thread and writes the file.

The files I tested with were plain m4b audio. So I used your Mp4File base class. I understand this still involves two separate threads for decoding and encoding frames.

I will have to keep this in mind when applying throttling policies in AAX Audio Converter for parallel processing of tracks.

> C is a pain, but I finally got the hang of it lol.

_I fully understand that. I was confronted with good old C the other day when porting 1990s 3rd party code from Win 32bit to Linux 64bit. Fortunately the originals were Unix, so no Win32 API dependencies there. Unfortunately, those programmers back then assumed 32bit pointers for their stuff. Oh dear. And syntax in C is significantly different from C++, I had to learn about structs again. I would never have found and resolved those issues without a proper debugger. I quite like the recently added CMake functionality within Visual Studio. It is far from perfect but its remote gdb capabilities are really nice._

**Reply by Mbucari** [reply]
> I will have to keep this in mind when applying throttling policies in AAX Audio Converter for parallel processing of tracks.

Some things to note.

- Reading from the input stream is I/O bound and is nearly free.
- Lossless decoding is I/O bound, so CPU utilization should always be less than 1 CPU thread.
- Decryption is nearly instantaneous and only runs on it's own thread because AAXClean's producer-consumer architecture. It's so inexpensive that you can discount it when estimating CPU usage.
- All of AAXClean's "threads" are tasks in the thread pool, so you may have many more threads than your system has CPU threads.
- I've generally found that:
  - Silence detection takes about 1/5 as much CPU time as decoding, so CPU utilization is ~1.2x CPU threads + I/O overhead.
  - Decoding takes about 1/10 as much CPU time as Mono Mp3 encoding, so CPU utilization is ~1.1x CPU threads + I/O overhead.
  - Stereo Mp3 encoding speed << Decoding speed, so CPU utilization is slightly more than 1 CPU thread + I/O overhead.
  - All AAC encoding speed << Decoding speed, so CPU utilization is slightly more than 1 CPU thread + I/O overhead.

Also, I modified the silence detection algorithm to use AVX2 CPU intrinsics. Previous versions used SSE4.1 which have generally been supported by Intel and AMD since ~2008. AVX2 is much newer and has been supported by Intel CPUs since about 2013 and AMD CUPs since about 2015.

**Reply by audiamus** [reply]
Very helpful. Thank you.

> producer-consumer architecture
_Oh, the good old design patterns_ 👍
